doc/src/manual/methods.md

```diff
@@ -629,4 +632,209 @@ function emptyfunc
+where (for example) you dispatch on the first argument and then call
+an internal method:
+
+```julia
+f(x::A, y) = _fA(x, y)
+f(x::B, y) = _fB(x, y)
+```
+
+Then the internal methods `_fA` and `_fB` can dispatch on `y` without
+concern about ambiguities with each other with respect to `x`.
+
+Be aware that this strategy has at least one major disadvantage: in
+many cases, it is not possible for users to further customize the
+behavior of `f` by defining further specializations of your exported
+function `f`. Instead, they have to define specializations for your
+internal methods `_fA` and `_fB`, and this blurs the lines between
+exported and internal methods.
+
+### Abstract containers and element types
+
+Where possible, try to avoid defining methods that dispatch on
+specific element types of abstract containers. For example,
+
+```julia
+-{T<:Date}(A::AbstractArray{T}, b::Date)
+```
+
+generates ambiguities for anyone who defines a method
+
+```julia
+-{T}(A::MyArrayType{T}, b::T)
+```
+
+The best approach is to avoid defining *either* of these methods:
+instead, rely on a generic method `-(A::AbstractArray, b)` and make
+sure this method is implemented with generic calls (like `similar` and
+`-`) that do the right thing for each container type and element type
+*separately*. This is just a more complex variant of the advice to
+[orthogonalize](@ref man-methods-orthogonalize) your methods.
+
+When this approach is not possible, it may be worth starting a
+discussion with other developers about resolving the ambiguity; just
+because one method was defined first does not necessarily mean that it
+can't be modified or eliminated.  As a last resort, one developer can
+define the "band-aid" method
+
+```julia
+-{T<:Date}(A::MyArrayType{T}, b::Date) = ...
+```
+
+that resolves the ambiguity by brute force.
+
+### Complex method "cascades" with default arguments
+
+If you are defining a method "cascade" that supplies defaults, be
+careful about dropping any arguments that correspond to potential
+defaults. For example, suppose you're writing a digital filtering
+algorithm and you have a method that handles the edges of the signal
+by applying padding:
+
+```julia
+function myfilter(A, kernel, ::Replicate)
+    Apadded = replicate_edges(A, size(kernel))
+    myfilter(Apadded, kernel)  # now perform the "real" computation
+end
+```
+
+This will run afoul of a method that supplies default padding:
+
+```julia
+myfilter(A, kernel) = myfilter(A, kernel, Replicate()) # replicate the edge by default
+```
+
+Together, these two methods generate an infinite recursion with `A` constantly growing bigger.
+
+The better design would be to define your call hierarchy like this:
+
+```julia
+struct NoPad end  # indicate that no padding is desired, or that it's already applied
+
+myfilter(A, kernel) = myfilter(A, kernel, Replicate())  # default boundary conditions
+
+function myfilter(A, kernel, ::Replicate)
+    Apadded = pad(A, kernel)
```

Comment:
would it make more sense for this to be the same padding function `replicate_edges(A, size(kernel))` as above?